sonarlint-vcs/src/main/java/org/sonarsource/sonarlint/core/vcs/GitUtils.java

```diff
@@ -0,0 +1,136 @@
+
+
+public class GitUtils {
+
+  private static final String BRANCH_REF_PREFIX = "refs/heads/";
+
+  private GitUtils() {
+    // util class
+  }
+
+  private static final SonarLintLogger LOG = SonarLintLogger.get();
+
+  public static List<String> getCommitNamesForRef(String branchName, Git git) {
+    var commitNames = new ArrayList<String>();
+    try {
+      var branchObjectId = git.getRepository().resolve(branchName);
+      if (branchObjectId == null) {
+        return commitNames;
+      }
+      var commits = git.log().add(branchObjectId).call();
+      for (RevCommit commit : commits) {
+        commitNames.add(commit.getName());
+      }
+    } catch (IncorrectObjectTypeException e) {
+      // do nothing, jgit will throw it if branchName can not be resolved to any commit
+    } catch (GitAPIException | IOException e) {
+      LOG.error("Couldn't fetch commits for branch " + branchName, e);
+    }
+    return commitNames;
+  }
+
+  @CheckForNull
+  public static Git getGitForDir(URI projectDirUri) {
+    try {
+      var builder = new FileRepositoryBuilder();
+      var gitUri = new URI(projectDirUri + "/.git");
+      var repository = builder.setGitDir(new File(gitUri)).setMustExist(true).build();
+      return new Git(repository);
+    } catch (IOException | URISyntaxException e) {
+      LOG.error("Couldn't access repository for path " + projectDirUri.getPath());
+    }
+    return null;
+  }
+
+
+  public static Optional<String> electSQBranchForLocalBranch(String branchName, Git git, Set<ServerBranch> serverCandidates) {
+    Set<String> serverCandidateNames = serverCandidates.stream().map(ServerBranch::getName).collect(Collectors.toSet());
+    Map<String, List<String>> commitsCache = buildCommitsCache(git);
+    if (commitsCache.isEmpty()) {
+      return Optional.empty();
+    }
+    List<String> commitNamesForBranch = getCommitNamesForRef(branchName, git);
+    Optional<String> mainBranchName = serverCandidates.stream().filter(ServerBranch::isMain).map(ServerBranch::getName).findFirst();
+    if (mainBranchName.isEmpty()) {
+      return Optional.empty();
+    }
+    List<String> listOfLocalCandidates;
+    for (String commitName : commitNamesForBranch) {
+      if (commitsCache.containsKey(commitName)) {
+        listOfLocalCandidates = commitsCache.get(commitName);
+        if (listOfLocalCandidates.contains(mainBranchName.get())) {
+          return mainBranchName;
+        }
+        for (String localCandidateName : listOfLocalCandidates) {
+          if (serverCandidateNames.contains(localCandidateName)) {
+            return Optional.of(localCandidateName);
+          }
+        }
+      }
+    }
+    return mainBranchName;
+  }
+
+  public static Map<String, List<String>> buildCommitsCache(Git git) {
+    List<Ref> refs;
+    try {
+      refs = git.getRepository().getRefDatabase().getRefs();
+    } catch (IOException e) {
+      LOG.error("Unable to build commits " + e);
```

Comment:
Better to pass the original exception as second argument, to preserve the stacktrace